metadata_mapper/mappers/mapper.py

```diff
@@ -37,62 +68,102 @@ def get_s3_api_response(self):
+    def to_dict(self) -> dict[str, Any]:
+        return self.mapped_metadata
 
+    def to_UCLDC(self) -> dict[str, Any]:
+        """
+        Maps source metadata to UCLDC format, saving result to
+        self.mapped_metadata.
 
-class Record(object):
-    def __init__(self, col_id, record):
-        self.collection_id = col_id
-        self.source_metadata = record
+        Returns: dict
+        """
+        self.mapped_metadata = {}
+
+        supermaps = [
+            super(c, self).UCLDC_map()
+            for c in list(reversed(type(self).__mro__))
+            if hasattr(super(c, self), "UCLDC_map")
+        ]
+        for map in supermaps:
+            self.mapped_metadata = {**self.mapped_metadata, **map}
+        self.mapped_metadata = {**self.mapped_metadata, **self.UCLDC_map()}
+
+        return self.mapped_metadata
+
+    def UCLDC_map(self) -> dict:
+        """
+        Defines mappings from source metadata to UCDLC that are specific
+        to this implementation.
+
+        All dicts returned by this method up the ancestor chain
+        are merged together to produce a final result. `is_show_at` and
+        `is_shown_by` are not fields from source metadata and are
+        always assumed to be computed, therefore they are included here
+        as maps to abstracted methods.
+        """
+        return {
+            "isShownAt": self.map_is_shown_at(),
+            "isShownBy": self.map_is_shown_by()
+        }
+
+    @abstractmethod
+    def map_is_shown_at(self) -> Union[str, None]:
+        pass
+
+    @abstractmethod
+    def map_is_shown_by(self) -> Union[str, None]:
+        pass
 
     # Mapper Helpers
+    def collate_plucked_values(self, values: list, pluck: str) -> list:
+        return [f[pluck] for f in values]
+
+    def collate_values(self, values):
+        collated = []
+        for value in values:
+            if not value:
+                continue
+
+            if isinstance(value, str):
+                collated.append(value)
+            else:
+                collated.extend(value)
+        return collated
+
     def collate_subfield(self, field, subfield):
+        """DEPRECATED: replace with `collate_plucked_values()` in mappers that use it"""
         return [f[subfield] for f in self.source_metadata.get(field, [])]
 
     def collate_fields(self, fieldlist):
-        ''' collate multiple field values into a single list '''
-        collated = []
-        for field in fieldlist:
-            value = self.source_metadata.get(field)
-            if value:
-                if isinstance(value, str):
-                    collated.append(value)
-                else:
-                    collated.extend(value)
+        """multiple field values into a single list
 
-        return collated
+        DEPRECATED: replace with `collate_values()` in mappers that use it
+        """
+        return self.collate_values([self.source_metadata.get(field) for field in fieldlist])
```

Comment:
Might be nice to pull in https://pypi.org/project/Python-Deprecated/ and mark these `@deprecated` so we don't forget them.

Reply:
Done